massa-protocol-worker-2/src/handlers/operation_handler/propagation.rs

```diff
@@ -0,0 +1,148 @@
+                    next_announce = std::time::Instant::now()
+                        .checked_add(self.config.operation_announcement_interval.to_duration())
+                        .expect("Can't init interval op propagation");
+                }
+                Err(RecvTimeoutError::Disconnected) => {
+                    return;
+                }
+            }
+        }
+    }
+
+    fn announce_ops(&mut self) {
+        // Quit if empty  to avoid iterating on nodes
+        if self.operations_to_announce.is_empty() {
+            return;
+        }
+        let operation_ids = mem::take(&mut self.operations_to_announce);
+        massa_trace!("protocol.protocol_worker.announce_ops.begin", {
+            "operation_ids": operation_ids
+        });
+        {
+            let mut cache_write = self.cache.write();
+            let peers: Vec<PeerId> = cache_write
+                .ops_known_by_peer
+                .iter()
+                .map(|(id, _)| id.clone())
+                .collect();
+            // Clean shared cache if peers do not exist anymore
+            for peer_id in peers {
+                if !self
+                    .active_connections
+                    .read()
+                    .connections
+                    .contains_key(&peer_id)
+                {
+                    cache_write.ops_known_by_peer.pop(&peer_id);
+                }
+            }
+            // Propagate to peers
+            for (peer_id, ops) in cache_write.ops_known_by_peer.iter_mut() {
+                let new_ops: Vec<OperationId> = operation_ids
+                    .iter()
+                    .filter(|id| !ops.contains(&id.prefix()))
+                    .copied()
+                    .collect();
+                if !new_ops.is_empty() {
+                    for id in &new_ops {
+                        ops.put(id.prefix(), ());
+                    }
+                    {
+                        let mut active_connections = self.active_connections.write();
+                        if let Some(connection) = active_connections.connections.get_mut(peer_id) {
+                            if let Err(err) = connection.send_channels.send(
+                                &self.operation_message_serializer,
+                                OperationMessage::OperationsAnnouncement(
+                                    new_ops.iter().map(|id| id.into_prefix()).collect(),
+                                )
+                                .into(),
+                                false,
+                            ) {
+                                debug!(
+                                    "could not send operation batch to node {}: {}",
+                                    peer_id, err
+                                );
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub fn start_propagation_thread(
```

Comment:
I really like this way of organizing things in the code !
A struct containing all the data the thread needs, a `run` method that contains the logic, and a constructor function that spawns the thread.
I think all the code spawning threads should be as close as possible to that organization, it makes it readable and very practical ! (That's maybe also my Python ex-dev POV)

Reply:
Yes I agree also ! We use it a lot around the code, you will be happy !